docs/architecture/adr-022-system-merge.md

```diff
@@ -0,0 +1,63 @@
+
+The next step migrates first the storage engine and then the syncing and fraud protocols (`go-header` and `go-fraud`) to `celestia-app`. This replaces the existing `BlockStore` and blocksync and evidence reactors. Streamlining the storage will require careful migration but will eliminate the duplication of storage and unify many of the data exchange protocols and the public APIs. Here, the consensus node can become more integrated with the `nodebuilder` package as it begins to be split up into a series of modules. This phase may also try to align configuration files and the general file directory.
+
+### Phase 3: Rewiring consensus, txpool and statesync
+
+The third phase consists of replacing the interface between the consensus, mempool and statesync protocols to Tendermint's p2p layer with LibP2P. This can be done in tandem with phase 2 and may be coupled with other refactors and performance improvements to those components. This will manifest itself as four new modules to `celestia-node`: consensus, mempool, statesync, and execution (a standalone Celestia state machine. This is somewhat similar to `app.App`).
+
+### Phase 4: Unification of Public APIs
+
+Lastly, legacy APIs like the RPC layer will be deprecated in favour of a unified public API. Every node should offer the same collection of endpoints and be capable of verifying all retrieved data.
+
+## Considerations
+
+This is a project of considerable duration (roughly 18 months) and thus will coincide with other feature developments. Careful coordination can be extremely helpful in reducing the engineering resources required to complete the work and avoid unnecessary complexity.
+
+- Phase 1 has no moving pieces it has to compete with and is relatively low risk.
+- Phase 2 needs to be aware of the introduction of pruning and partial nodes. It also needs to be aware of potential improvements to the header verification algorithm. Phase 2 will also require a lazy migration of data types. Both legacy and new syncing and fraud protocols will need to be run side by side.
+- Phase 3 will likely coincide with bandwidth and latency improvements to both the consensus and mempool protocols such as: [Compact Blocks](https://github.com/celestiaorg/celestia-core/issues/883), Tx Client, Intermediate Blocks, Transaction Sharding (i.e. Narwhal), [BlobTx Pipe](https://github.com/celestiaorg/celestia-app/issues/2297), and Consensus Pipelining. Statesync will likely remain unchanged and be the easiest to support. Giving rolling upgrades, there will be one version that will support both Tendermint P2P and LibP2P for all components.
+
+### Risk Mitigation
+
+Replacing the networking layer is a significant undertaking and vulnerable to a number of risks. Each component will need to be benchmared to ensure no performance degradation and be sufficiently stress tested. Having both versions present, while adding complexity, has the advantage of enabling the option to fallback in case of unexpected failure.
```

Comment:
```suggestion
Replacing the networking layer is a significant undertaking and vulnerable to a number of risks. Each component will need to be benchmarked to ensure no performance degradation and be sufficiently stress tested. Having both versions present, while adding complexity, has the advantage of enabling the option to fallback in case of unexpected failure.
```